services/src/main/java/org/keycloak/services/x509/Rfc9440ClientCertificateLookup.java

```diff
@@ -0,0 +1,186 @@
+import org.keycloak.common.util.DerUtils;
+import org.keycloak.http.HttpRequest;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.GeneralSecurityException;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * The provider allows to extract a client certificate forwarded
+ * to the keycloak middleware configured behind a reverse proxy that is
+ * compliant with RFC 9440.
+ *
+ * @author <a href="mailto:seiferma.dev+kc@gmail.com">Stephan Seifermann</a>
+ * @version $Revision: 1 $
+ * @since 12/30/2024
+ */
+public class Rfc9440ClientCertificateLookup implements X509ClientCertificateLookup {
+
+    public static class RfcViolationException extends Exception {
+        public RfcViolationException(String rfc, String section, String details, Throwable cause) {
+            super("Violation of RFC " + rfc + " (see section " + section + "): " + details, cause);
+        }
+    }
+
+    public static class Rfc9440ViolationException extends RfcViolationException {
+        public Rfc9440ViolationException(String section, String details) {
+            this(section, details, null);
+        }
+        public Rfc9440ViolationException(String section, String details, Throwable cause) {
+            super("9440", section, details, cause);
+        }
+    }
+
+    public static class Rfc8941ViolationException extends RfcViolationException {
+        public Rfc8941ViolationException(String section, String details) {
+            this(section, details, null);
+        }
+        public Rfc8941ViolationException(String section, String details, Throwable cause) {
+            super("8941", section, details, cause);
+        }
+    }
+
+    private static final Logger log = Logger.getLogger(Rfc9440ClientCertificateLookup.class);
+    protected final String sslClientCertHttpHeader;
+    protected final String sslCertChainHttpHeader;
+    protected final int certificateChainLength;
+
+    public Rfc9440ClientCertificateLookup(String sslClientCertHttpHeader,
+                                          String sslCertChainHttpHeader,
+                                          int certificateChainLength) {
+        this.sslClientCertHttpHeader = Optional.ofNullable(sslClientCertHttpHeader)
+                .filter(s -> !s.isBlank())
+                .orElseThrow(() ->  new IllegalArgumentException("sslClientCertHttpHeader"));
+
+        this.sslCertChainHttpHeader = Optional.ofNullable(sslCertChainHttpHeader)
+                .filter(s -> !s.isBlank())
+                .orElseThrow(() ->  new IllegalArgumentException("sslCertChainHttpHeader"));
+
+        this.certificateChainLength = certificateChainLength;
+    }
+
+    @Override
+    public X509Certificate[] getCertificateChain(HttpRequest httpRequest) throws GeneralSecurityException {
```

Comment:
This will need to check if the header is trusted - https://github.com/keycloak/keycloak/blob/a72d15b8572926bc460181f9338f7bf65f4ac04c/services/src/main/java/org/keycloak/services/x509/AbstractClientCertificateFromHttpHeadersLookup.java#L120

Reply:
I rebased my branch and integrated the check. Thanks for the hint!